File: swift/YatzyTests.swift
import XCTest
@testable import Yatzy_Refactoring_Kata

class YatzyTests: XCTestCase {
    
    func testChanceScoresSumOfAllDice() {
        let expected = 15
        let actual = Yatzy.chance(d1: 2, d2: 3, d3: 4, d4: 5, d5: 1)
        XCTAssertEqual(expected, actual)
        XCTAssertEqual(16, Yatzy.chance(d1: 3, d2: 3, d3: 4, d4: 5, d5: 1))
    }
    
    func testYatzyScores50() {
        let expected = 50
        let actual = Yatzy.yatzy(dice: 4, 4, 4, 4, 4)
        XCTAssertEqual(expected, actual)
        XCTAssertEqual(50, Yatzy.yatzy(dice: 6, 6, 6, 6, 6))
        XCTAssertEqual(0, Yatzy.yatzy(dice: 6, 6, 6, 6, 3))
    }
    
    func test1s() {
        XCTAssertTrue(Yatzy.ones(d1: 1, d2: 2, d3: 3, d4: 4, d5: 5) == 1)
        XCTAssertEqual(2, Yatzy.ones(d1: 1, d2: 2, d3: 1, d4: 4, d5: 5))
        XCTAssertEqual(0, Yatzy.ones(d1: 6, d2: 2, d3: 2, d4: 4, d5: 5))
        XCTAssertEqual(4, Yatzy.ones(d1: 1, d2: 2, d3: 1, d4: 1, d5: 1))
    }
    
    func test2s() {
        XCTAssertEqual(4, Yatzy.twos(d1: 1, d2: 2, d3: 3, d4: 2, d5: 6))
        XCTAssertEqual(10, Yatzy.twos(d1: 2, d2: 2, d3: 2, d4: 2, d5: 2))
    }

    func test3s() {
        XCTAssertEqual(6, Yatzy.threes(d1: 1, d2: 2, d3: 3, d4: 2, d5: 3))
        XCTAssertEqual(12, Yatzy.threes(d1: 2, d2: 3, d3: 3, d4: 3, d5: 3))
    }
    
    func test4s() {
        XCTAssertEqual(12, Yatzy(d1: 4, d2: 4, d3: 4, d4: 5, _5: 5).fours())
        XCTAssertEqual(8, Yatzy(d1: 4, d2: 4, d3: 5, d4: 5, _5: 5).fours())
        XCTAssertEqual(4, Yatzy(d1: 4, d2: 5, d3: 5, d4: 5, _5: 5).fours())
    }

    func test5s() {
        XCTAssertEqual(10, Yatzy(d1: 4, d2: 4, d3: 4, d4: 5, _5: 5).fives())
        XCTAssertEqual(15, Yatzy(d1: 4, d2: 4, d3: 5, d4: 5, _5: 5).fives())
        XCTAssertEqual(20, Yatzy(d1: 4, d2: 5, d3: 5, d4: 5, _5: 5).fives())
    }
    
    func test6s() {
        XCTAssertEqual(0, Yatzy(d1: 4, d2: 4, d3: 4, d4: 5, _5: 5).sixes())
        XCTAssertEqual(6, Yatzy(d1: 4, d2: 4, d3: 6, d4: 5, _5: 5).sixes())
        XCTAssertEqual(18, Yatzy(d1: 6, d2: 5, d3: 6, d4: 6, _5: 5).sixes())
    }
    
    func testOnePair() {
        XCTAssertEqual(6, Yatzy.scorePair(d1: 3, d2: 4, d3: 3, d4: 5, d5: 6))
        XCTAssertEqual(10, Yatzy.scorePair(d1: 5, d2: 3, d3: 3, d4: 3, d5: 5))
        XCTAssertEqual(12, Yatzy.scorePair(d1: 5, d2: 3, d3: 6, d4: 6, d5: 5))
    }
    
    func testTwoPairs() {
        XCTAssertEqual(16, Yatzy.twoPairs(d1: 3, d2: 3, d3: 5, d4: 4, d5: 5))
        XCTAssertEqual(16, Yatzy.twoPairs(d1: 3, d2: 3, d3: 5, d4: 5, d5: 5))
    }
    
    func testThreeOfAKind() {
        XCTAssertEqual(9, Yatzy.threeOfAKind(d1: 3, d2: 3, d3: 3, d4: 4, d5: 5))
        XCTAssertEqual(15, Yatzy.threeOfAKind(d1: 5, d2: 3, d3: 5, d4: 4, d5: 5))
        XCTAssertEqual(9, Yatzy.threeOfAKind(d1: 3, d2: 3, d3: 3, d4: 3, d5: 5))
    }
    
    func testFourOfAKind() {
        XCTAssertEqual(12, Yatzy.fourOfAKind(_1: 3, _2: 3, d3: 3, d4: 3, d5: 5))
        XCTAssertEqual(20, Yatzy.fourOfAKind(_1: 5, _2: 5, d3: 5, d4: 4, d5: 5))
        XCTAssertEqual(9, Yatzy.threeOfAKind(d1: 3, d2: 3, d3: 3, d4: 3, d5: 3))
    }
    
    func testSmallStraight() {
        XCTAssertEqual(15, Yatzy.smallStraight(d1: 1, d2: 2, d3: 3, d4: 4, d5: 5))
        XCTAssertEqual(15, Yatzy.smallStraight(d1: 2, d2: 3, d3: 4, d4: 5, d5: 1))
        XCTAssertEqual(0, Yatzy.smallStraight(d1: 1, d2: 2, d3: 2, d4: 4, d5: 5))
    }
    
    func testLargeStraight() {
        XCTAssertEqual(20, Yatzy.largeStraight(d1: 6, d2: 2, d3: 3, d4: 4, d5: 5))
        XCTAssertEqual(20, Yatzy.largeStraight(d1: 2, d2: 3, d3: 4, d4: 5, d5: 6))
        XCTAssertEqual(0, Yatzy.largeStraight(d1: 1, d2: 2, d3: 3, d4: 4, d5: 5))
    }
    
    func testFullHouse() {
        XCTAssertEqual(18, Yatzy.fullHouse(d1: 6, d2: 2, d3: 2, d4: 2, d5: 6))
        XCTAssertEqual(0, Yatzy.fullHouse(d1: 2, d2: 3, d3: 4, d4: 5, d5: 6))
    }
}

File: scala/src/test/scala/YatzyTests.scala
import Yatzy.Yatzy
import org.scalatest.funsuite.AnyFunSuite
import org.scalatest.matchers.should.Matchers

class YatzyTests extends AnyFunSuite with Matchers {
  test("Chance scores sum of all dice") {
    val expected = 15
    val actual = Yatzy.chance(2, 3, 4, 5, 1)

    assert(actual == expected)
    assert(Yatzy.chance(3, 3, 4, 5, 1) == 16)
  }

  test("test_1s") {
    assert(Yatzy.ones(1, 2, 3, 4, 5) == 1)
    assert(Yatzy.ones(1, 2, 1, 4, 5) == 2)
    assert(Yatzy.ones(6, 2, 2, 4, 5) == 0)
    assert(Yatzy.ones(1, 2, 1, 1, 1) == 4)
  }

  test("test_2s") {
    assert(Yatzy.twos(1, 2, 3, 2, 6) == 4)
    assert(Yatzy.twos(2, 2, 2, 2, 2) == 10)
  }

  test("test_3s") {
    assert(Yatzy.threes(1, 2, 3, 2, 3) == 6)
    assert(Yatzy.threes(2, 3, 3, 3, 3) == 12)
  }

  test("test_5s") {
    assert(new Yatzy(4, 4, 4, 5, 5).fives() == 10)
    assert(new Yatzy(4, 4, 5, 5, 5).fives() == 15)
    assert(new Yatzy(4, 5, 5, 5, 5).fives() == 20)
  }

  test("four of a knd") {
    assert(Yatzy.fourOfAKind(3, 3, 3, 3, 5) == 12)
    assert(Yatzy.fourOfAKind(5, 5, 5, 4, 5) == 20)
    assert(Yatzy.fourOfAKind(3, 3, 3, 3, 3) == 12)
  }

  test("test_4s") {
    assert(new Yatzy(4, 4, 4, 5, 5).fours() == 12)
    assert(new Yatzy(4, 4, 5, 5, 5).fours() == 8)
    assert(new Yatzy(4, 5, 5, 5, 5).fours() == 4)
  }

  test("fullHouse") {
    assert(Yatzy.fullHouse(6, 2, 2, 2, 6) == 18)
    assert(Yatzy.fullHouse(2, 3, 4, 5, 6) == 0)
  }

  test("largeStraight") {
    assert(Yatzy.largeStraight(6, 2, 3, 4, 5) == 20)
    assert(Yatzy.largeStraight(2, 3, 4, 5, 6) == 20)
    assert(Yatzy.largeStraight(1, 2, 2, 4, 5) == 0)
  }

  test("one Pair") {
    assert(Yatzy.scorePair(3, 4, 3, 5, 6) == 6)
    assert(Yatzy.scorePair(5, 3, 3, 3, 5) == 10)
    assert(Yatzy.scorePair(5, 3, 6, 6, 5) == 12)
  }

  test("sixes") {
    assert(new Yatzy(4, 4, 4, 5, 5).Sixes() == 0)
    assert(new Yatzy(4, 4, 6, 5, 5).Sixes() == 6)
    assert(new Yatzy(6, 5, 6, 6, 5).Sixes() == 18)
  }

  test("smallStraight") {
    assert(Yatzy.smallStraight(1, 2, 3, 4, 5) == 15)
    assert(Yatzy.smallStraight(2, 3, 4, 5, 1) == 15)
    assert(Yatzy.smallStraight(1, 2, 2, 4, 5) == 0)
  }

  test("three_of_a_kind") {
    assert(Yatzy.threeOfAKind(3, 3, 3, 4, 5) == 9)
    assert(Yatzy.threeOfAKind(5, 3, 5, 4, 5) == 15)
    assert(Yatzy.threeOfAKind(3, 3, 3, 3, 5) == 9)
  }

  test("two pair") {
    assert(Yatzy.twoPair(3, 3, 5, 4, 5) == 16)
    assert(Yatzy.twoPair(3, 3, 5, 5, 5) == 16)
  }

  test("Yatzy scores 50") {
    val expected = 50
    val actual = Yatzy.Yatzy(4, 4, 4, 4, 4)
    assert(actual == expected)
    assert(Yatzy.Yatzy(6, 6, 6, 6, 6) == 50)
    assert(Yatzy.Yatzy(6, 6, 6, 6, 3) == 0)
  }
}
